### ISSUE TITLE
release_1_3 clientset should use a serializer instead of a codec

### ISSUE THREAD

**Issue by caesarxuchao** [kubernetes/kubernetes#25589]
#24789 makes the release_1_3 clientset use a codec that does conversion. A versioned clientset should use a serializer that doesn't do any conversion, because its methods return versioned objects, not internal objects.

There is some discussion in #24789: https://github.com/kubernetes/kubernetes/pull/24789/files/b4c83022e382f60dc9cdbcf65a841c10fdac17a5..3aadafd4119dd1c59218e2a4ee032d64975b5d7f#r62389615

@nikhiljindal @smarterclayton @wojtek-t @lavalamp @kubernetes/sig-api-machinery

**Comment by nikhiljindal** [comment]
cc @mfanjie who also ran into this while using using versioned federation_clientset. Its returning unversioned objects when it should return versioned objects.

**Comment by smarterclayton** [comment]
The client set should use the serializer, but today it must take the responsibility of setting the correct GroupVersionKind on the object prior to invoking the serializer.  It may be necessary to create a Serialization wrapper or a non-converting Codec that do exactly one thing - look up the object to serialize's type in an ObjectTyper and copy it to the GroupVersionKind() a la setTargetKind in runtime/scheme.go.  

Basically:

```
type DirectCodec struct {
  Serializer
  ObjectTyper
}

func (c DirectCodec) EncodeToStream(obj runtime.Object, ...) error {
  gvk, err := c.ObjectTyper.ObjectKind(obj)
  ...
  obj.SetGroupVersionKind(gvk)
  return c.Serializer.Encode(obj, ...)
}
```

**Comment by smarterclayton** [comment]
It can ignore overrides for now, but in theory overrides should influence what kind is set.

**Comment by caesarxuchao** [comment]
I'm working on this and will let you review.
